src/machine/machine-dbus.c

```diff
@@ -621,24 +621,39 @@ int bus_machine_method_open_shell(sd_bus_message *message, void *userdata, sd_bu
         if (r < 0)
                 return r;
         user = empty_to_null(user);
-        if (isempty(path))
-                path = "/bin/sh";
-        if (!path_is_absolute(path))
-                return sd_bus_error_setf(error, SD_BUS_ERROR_INVALID_ARGS, "Specified path '%s' is not absolute", path);
-
-        r = sd_bus_message_read_strv(message, &args);
+        r = sd_bus_message_read_strv(message, &v);
         if (r < 0)
                 return r;
-        if (strv_isempty(args)) {
-                args = strv_free(args);
-
-                args = strv_new(path, NULL);
+        if (isempty(path)) {
+                path = "/bin/sh";
+                args = new0(char*, 3 + 1);
                 if (!args)
                         return -ENOMEM;
-
-                args[0][0] = '-'; /* Tell /bin/sh that this shall be a login shell */
+                args[0] = strdup("sh");
+                if (!args[0])
+                        return -ENOMEM;
+                args[1] = strdup("-c");
+                if (!args[1])
+                        return -ENOMEM;
+                r = asprintf(&args[2], /* This only forks one process, getent, and checks for its existence. */
+                             "shell=$(getent passwd %s 2>/dev/null | { IFS=: read _ _ _ _ _ _ x; echo \"$x\"; });\n"\
+                             "exec \"${shell:-/bin/sh}\" -l" /* short for --login */,
```

Comment:
I'm not sure all shells understand the `-l` argument, at least I didn't find it documented anywhere.

Reply:
Do you know how to set argv[0][0] in shell? exec -a is a bashism.

-l is supported (and documented) by bash dash and zsh, and supported by busybox's ash.